jobs: TestRetriesWithExponentialBackoff failed

jobs.TestRetriesWithExponentialBackoff [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4034895&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4034895&tab=artifacts#/) on master @ [e3be114fc6fe8eda07aafbc7ee43cc452456f270](https://github.com/cockroachdb/cockroach/commits/e3be114fc6fe8eda07aafbc7ee43cc452456f270):


```
=== RUN   TestRetriesWithExponentialBackoff
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/122da43dba5ccb08c66085e52c133b84/logTestRetriesWithExponentialBackoff3622445839
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestRetriesWithExponentialBackoff
    registry_test.go:746: -- test log scope end --
--- FAIL: TestRetriesWithExponentialBackoff (7.18s)
=== RUN   TestRetriesWithExponentialBackoff/pause_running
    registry_test.go:596: 
        	Error Trace:	registry_test.go:596
        	            				registry_test.go:652
        	Error:      	Not equal: 
        	            	expected: 34
        	            	actual  : 35
        	Test:       	TestRetriesWithExponentialBackoff/pause_running
        	Messages:   	unexpected number of jobs resumed in retry 33
    --- FAIL: TestRetriesWithExponentialBackoff/pause_running (1.20s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRetriesWithExponentialBackoff.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-12062

jobs.TestRetriesWithExponentialBackoff [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4033710&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4033710&tab=artifacts#/) on master @ [e3be114fc6fe8eda07aafbc7ee43cc452456f270](https://github.com/cockroachdb/cockroach/commits/e3be114fc6fe8eda07aafbc7ee43cc452456f270):


```
=== RUN   TestRetriesWithExponentialBackoff
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRetriesWithExponentialBackoff3998464948
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestRetriesWithExponentialBackoff
    registry_test.go:746: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRetriesWithExponentialBackoff3998464948
--- FAIL: TestRetriesWithExponentialBackoff (10.17s)
=== RUN   TestRetriesWithExponentialBackoff/pause_running
    registry_test.go:596: 
        	Error Trace:	registry_test.go:596
        	            				registry_test.go:652
        	Error:      	Not equal: 
        	            	expected: 48
        	            	actual  : 49
        	Test:       	TestRetriesWithExponentialBackoff/pause_running
        	Messages:   	unexpected number of jobs resumed in retry 47
    --- FAIL: TestRetriesWithExponentialBackoff/pause_running (2.34s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRetriesWithExponentialBackoff.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRetriesWithExponentialBackoff [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052653&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052653&tab=artifacts#/) on master @ [3401e4d902953f7684aebb819090f586c9acfec9](https://github.com/cockroachdb/cockroach/commits/3401e4d902953f7684aebb819090f586c9acfec9):


```
=== RUN   TestRetriesWithExponentialBackoff
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/122da43dba5ccb08c66085e52c133b84/logTestRetriesWithExponentialBackoff3564845755
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestRetriesWithExponentialBackoff
    registry_test.go:746: -- test log scope end --
--- FAIL: TestRetriesWithExponentialBackoff (9.45s)
=== RUN   TestRetriesWithExponentialBackoff/pause_reverting
    registry_test.go:596: 
        	Error Trace:	registry_test.go:596
        	            				registry_test.go:733
        	Error:      	Not equal: 
        	            	expected: 35
        	            	actual  : 36
        	Test:       	TestRetriesWithExponentialBackoff/pause_reverting
        	Messages:   	unexpected number of jobs resumed in retry 33
    --- FAIL: TestRetriesWithExponentialBackoff/pause_reverting (1.70s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRetriesWithExponentialBackoff.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRetriesWithExponentialBackoff [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052654&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052654&tab=artifacts#/) on master @ [3401e4d902953f7684aebb819090f586c9acfec9](https://github.com/cockroachdb/cockroach/commits/3401e4d902953f7684aebb819090f586c9acfec9):


```
=== RUN   TestRetriesWithExponentialBackoff
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/122da43dba5ccb08c66085e52c133b84/logTestRetriesWithExponentialBackoff2922246898
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestRetriesWithExponentialBackoff
    registry_test.go:746: -- test log scope end --
--- FAIL: TestRetriesWithExponentialBackoff (34.51s)
=== RUN   TestRetriesWithExponentialBackoff/pause_running
    registry_test.go:596: 
        	Error Trace:	registry_test.go:596
        	            				registry_test.go:652
        	Error:      	Not equal: 
        	            	expected: 36
        	            	actual  : 37
        	Test:       	TestRetriesWithExponentialBackoff/pause_running
        	Messages:   	unexpected number of jobs resumed in retry 35
    --- FAIL: TestRetriesWithExponentialBackoff/pause_running (6.36s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRetriesWithExponentialBackoff.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRetriesWithExponentialBackoff [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052495&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052495&tab=artifacts#/) on master @ [3401e4d902953f7684aebb819090f586c9acfec9](https://github.com/cockroachdb/cockroach/commits/3401e4d902953f7684aebb819090f586c9acfec9):


```
=== RUN   TestRetriesWithExponentialBackoff
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRetriesWithExponentialBackoff454964543
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestRetriesWithExponentialBackoff
    registry_test.go:746: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRetriesWithExponentialBackoff454964543
--- FAIL: TestRetriesWithExponentialBackoff (9.14s)
=== RUN   TestRetriesWithExponentialBackoff/pause_reverting
    registry_test.go:596: 
        	Error Trace:	registry_test.go:596
        	            				registry_test.go:733
        	Error:      	Not equal: 
        	            	expected: 23
        	            	actual  : 24
        	Test:       	TestRetriesWithExponentialBackoff/pause_reverting
        	Messages:   	unexpected number of jobs resumed in retry 21
    --- FAIL: TestRetriesWithExponentialBackoff/pause_reverting (1.18s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRetriesWithExponentialBackoff.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

jobs.TestRetriesWithExponentialBackoff [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052496&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4052496&tab=artifacts#/) on master @ [3401e4d902953f7684aebb819090f586c9acfec9](https://github.com/cockroachdb/cockroach/commits/3401e4d902953f7684aebb819090f586c9acfec9):


```
=== RUN   TestRetriesWithExponentialBackoff
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRetriesWithExponentialBackoff881193565
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestRetriesWithExponentialBackoff
    registry_test.go:746: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRetriesWithExponentialBackoff881193565
--- FAIL: TestRetriesWithExponentialBackoff (29.81s)
=== RUN   TestRetriesWithExponentialBackoff/pause_running
    registry_test.go:596: 
        	Error Trace:	registry_test.go:596
        	            				registry_test.go:652
        	Error:      	Not equal: 
        	            	expected: 11
        	            	actual  : 12
        	Test:       	TestRetriesWithExponentialBackoff/pause_running
        	Messages:   	unexpected number of jobs resumed in retry 10
    --- FAIL: TestRetriesWithExponentialBackoff/pause_running (2.36s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRetriesWithExponentialBackoff.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

I ran into this on 21.2 branch (80478cbb0341923a4cf22a946b8e0c469e1428d0) - it's quickly reproduced with `make stress`:
```
=== RUN   TestRetriesWithExponentialBackoff/pause_running
    registry_test.go:594: 
        	Error Trace:	registry_test.go:594
        	            				registry_test.go:650
        	Error:      	Not equal: 
        	            	expected: 8
        	            	actual  : 9
        	Test:       	TestRetriesWithExponentialBackoff/pause_running
        	Messages:   	unexpected number of jobs resumed in retry 7
panic: send on closed channel [recovered]
	panic: send on closed channel

goroutine 40855 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc002c13500, {0x5022810, 0xc0024f9cc0})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0xaa
panic({0x3797ea0, 0x4f6b670})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/jobs.TestRetriesWithExponentialBackoff.func6.3.1({0xc003eb7368, 0xc00240d5f0})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry_test.go:533 +0x8b
github.com/cockroachdb/cockroach/pkg/jobs.FakeResumer.Resume({0xc005670640, 0xc005670650, 0x0, 0x0, 0x0}, {0x50228b8, 0xc0015a7d10}, {0xc001c8e4b0, 0x0})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/helpers_test.go:37 +0x50
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0x50228b8, {0x4fcc810, 0xc0015a7b60}, {0x50228b8, 0xc0015a7d10}, {0x3a8ae40, 0xc00356e670}, 0xc00240d8d8)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1200 +0xba
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc00394b600, {0x50228b8, 0xc0015a7cb0}, {0x3a8ae40, 0xc00356e670}, {0x4fcc810, 0xc0015a7b60}, 0xc003eb7340, {0xc00051a718, 0x7}, ...)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1201 +0x657
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc00394b600, {0x5022810, 0xc0024779c0}, {0x4fcc810, 0xc0015a7b60}, 0xc003eb7340, {0xc00051a718, 0x7}, {0xc005094e88, 0x16})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:421 +0x551
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1({0x50228b8, 0xc0012c5530})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:342 +0xfa
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x123
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x387


ERROR: exit status 2

24 runs completed, 1 failures, over 1m18s
```
I will skip the test on 21.2 branch too.

--- a/pkg/jobs/registry_test.go
+++ b/pkg/jobs/registry_test.go
@@ -733,14 +733,24 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 						return nil
 					}
 					bti.resumeCh <- struct{}{}
-					return <-bti.errCh
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					case err := <-bti.errCh:
+						return err
+					}
 				},
 				FailOrCancel: func(ctx context.Context) error {
 					if bti.done.Load().(bool) {
 						return nil
 					}
 					bti.failOrCancelCh <- struct{}{}
-					return <-bti.errCh
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					case err := <-bti.errCh:
+						return err
+					}
 				},
 			}
 		}, UsesTenantCostControl)
@@ -793,7 +803,14 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 			expectedResumed++
 			retryCnt++
 			// Validate that the job is resumed only once.
-			require.Equal(t, expectedResumed, bti.resumed.Count(), "unexpected number of jobs resumed in retry %d", i)
+			//
+			// For tests that immediately resume, we can't make this
+			// assertion because the waitFn above might have already
+			// put the job into the state where it will be retried
+			// and by definition it is not going to wait to retry.
+			if !bti.expectImmediateRetry {
+				require.Equal(t, expectedResumed, bti.resumed.Count(), "unexpected number of jobs resumed in retry (post waitFn) %d", i)
+			}
 			lastRun = bti.clock.Now()
 		}
 		bti.done.Store(true)
@@ -839,7 +856,6 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 	t.Run("pause running", func(t *testing.T) {
 		ctx := context.Background()
 		bti := BackoffTestInfra{expectImmediateRetry: true}
-		skip.WithIssue(t, 74399)
 		bti.afterJobStateMachineKnob = func() {
 			if bti.done.Load().(bool) {
 				return
@@ -856,7 +872,6 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 			<-bti.resumeCh
 			insqlDB := bti.s.InternalDB().(isql.DB)
 			pauseOrCancelJob(t, ctx, insqlDB, bti.registry, jobID, pause)
-			bti.errCh <- nil
 			<-bti.transitionCh
 			waitUntilStatus(t, bti.tdb, jobID, StatusPaused)
 			require.NoError(t, bti.registry.Unpause(ctx, nil, jobID))
@@ -920,7 +935,6 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 	t.Run("pause reverting", func(t *testing.T) {
 		ctx := context.Background()
 		bti := BackoffTestInfra{expectImmediateRetry: true}
-		skip.WithIssue(t, 74399)
 
 		bti.afterJobStateMachineKnob = func() {
 			if bti.done.Load().(bool) {
@@ -945,11 +959,6 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 		runTest(t, jobID, retryCnt, expectedResumed, lastRun, &bti, func(_ int64) {
 			<-bti.failOrCancelCh
 			pauseOrCancelJob(t, ctx, bti.idb, bti.registry, jobID, pause)
-			// We have to return error here because, otherwise, the job will be marked as
-			// failed regardless of the fact that it is currently pause-requested in the
-			// jobs table. This is because we currently do not check the current status
-			// of a job before marking it as failed.
-			bti.errCh <- MarkAsRetryJobError(errors.New("injecting error in reverting state to retry"))
 			<-bti.transitionCh
 			waitUntilStatus(t, bti.tdb, jobID, StatusPaused)
 			require.NoError(t, bti.registry.Unpause(ctx, nil, jobID))

